### PR TITLE
Split package management into daemon and plugins components

### PR DESCRIPTION
**Pull request by gcoxmoz**
#### Pull Request (PR) description
This PR splits the `package_name` and `manage_package` parameters into two extra pairs of parameters, one for the daemon and one for the plugins.  This allows more granular control of the deployments.  In the use case of "the puppet module controls the daemon package but not the plugins package", without this PR you must replicate much of the os-based choices of `manifests/params.pp` outside of the module in order to ensure that you get the right package names, when a much cleaner path is to simply capture that information in the module itself.

This is somewhat clumsy in that it introduces a duplication in the API, where you could end up with two different code paths controlling the same packages, so I'm quite open to changing patch 2 to be 'better' in terms of what's enabled/disabled, or if `manage_package` is changed to false in preparation for deprecation, or whatever seems most sane.

### CODE REVIEW

File: manifests/install.pp [hunk 1]
```diff
@@ -4,9 +4,27 @@
 class nrpe::install
 {
   if $nrpe::manage_package {
-    package { $nrpe::package_name:
-      ensure   => installed,
-      provider => $nrpe::provider,
-    }
+    ensure_packages($nrpe::package_name,
```
Comment (baurmatt) [inline manifests/install.pp]: I would prefer to manage those packages with a "normal" `package` resource as this is the main module for this. If you have some special modules which might do the same, I think they should use a `ensure_packages`.

**Comment by baurmatt** [comment]
Hey @gcoxmoz, thanks for your PR! :) Can you please take a look at my comment and the failing tests?

File: manifests/install.pp (hunk 1 above)
Comment (gcoxmoz) [reply manifests/install.pp]: So, I can't change it the way you said without a bigger architectural difference, which I fear might make this a harder transition, and this was my compromise.

Someone upgrading to this patch who has set `package_name` expects their list to work.  With this patch, they'd be 'double included', and it'd just work.  And I could set options for myself and be happy.

To use "normal" `package`, we'd have to split all package requirements up so there's no overlap, because otherwise `package` will complain when we include packages during `if $nrpe::manage_package` and then for `if $nrpe::manage_package_daemon`.

To do otherwise, we'd have to flip the defaults to false for my patch, which I could be fine with.  Basically make it an exclusive-or situation. (`$nrpe::manage_package` XOR (if `$nrpe::manage_package_daemon` AND `if $nrpe::manage_package_plugins`)).  I could do that, it'd just mean reversing the defaults on the manage flags.

The other options I see would be major-version-breaking:
- do `if $nrpe::manage_package or $nrpe::manage_package_daemon { package {$nrpe::package_name_daemon} } ` - this would mean `package_name` never gets used, breaking expectations
- drop `if $nrpe::manage_package` completely in favor of my method.

Basically, I agree, using this module, 'it' should control the package (if you let it), which would mean it should be `package`.  I can do flag reversals.  If it's okay to make this a major version change instead of a first step towards deprecation, I can patch it, drop `ensure_packages` back to `package`, and there we go, but that's a hard cut.  Basically, I'm open for whichever seems most sane here.  Just saying, I didn't go for `ensure_packages` willy-nilly. :)

**Comment by gcoxmoz** [comment]
It looks like [tests](https://travis-ci.org/voxpupuli/puppet-nrpe/pull_requests?utm_medium=notification&utm_source=github_status) have been failing for a few months, in the same manner.  I'm not sure this PR is failing because of anything I've done.

File: manifests/install.pp (hunk 1 above)
Comment (baurmatt) [reply manifests/install.pp]: Hey @gcoxmoz, thanks for you answer! :) How about doing the following (and `manage_package`/`package_name`)  in `install.pp` based on the values from `init.pp`:

```puppet
  if $nrpe_packages_daemon and $nrpe_packages_plugins {
    $nrpe_packages = [ $nrpe_packages_daemon ] + $nrpe_packages_plugins
  }
```

This would then allow us to only do 1 package resource definition:

```puppet
package { $all_package_which_should_be_installed:
  ensure   => installed,
  provider => $nrpe::provider,
}
```

**Comment by baurmatt** [comment]
> It looks like tests have been failing for a few months, in the same manner. I'm not sure this PR is failing because of anything I've done.

You're right, sorry! :)

**Review by llowder — COMMENTED** [review]
I see no red flags, but I'll wait for someone more familiar with the tech to review.
